proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+
+```csharp
+((int x, int y, int z)?, int t)? SpaceTime()
+{
+    return ((1,2,3), 7);  // valid, implicit nullable conversion
+}
+```
+
+## Additions to [Expressions](../../spec/expressions.md)
+
+### Overload resolution and tuples with no natural types
+
+> The following text is added after after the bullet list in [Exactly matching expressions](../../spec/expressions.md#Exactly-matching-expressions):
+
+The exact match rule for tuple expressions is based on the natural types of the constituent tuple arguments. The rule is mutually recursive with respect to other containing or contained expressions not in a possession of a natural type.
+
+### Deconstruction expressions
+
+> This section should be added at the end of the [Expressions](../../spec/expressions.md) chapter.
+
+Tuple deconstruction expressions receive and "split out" a tuple:
+
+```csharp
+(var sum, var count) = Tally(myValues); // deconstruct result
+Console.WriteLine($"Sum: {sum}, count: {count}");  
+```
+
+The `_` wildcard indicates that the one or more of the tuple fields are discarded:
+
+```csharp
+(var sum, _) = Tally(myValues); // deconstruct result
+Console.WriteLine($"Sum: {sum}, count was ignored");  
+```
+
+Any object may be deconstructed by providing an accessible `Deconstruct` method, either as a member or as an extension method. A `Deconstruct` method converts an object to a set of discrete values. The Deconstruct method "returns" the component values by use of individual `out` parameters. Deconstruct is overloadable.
+
+The deconstructor pattern could be implemented as a member method, or an extension method:
+
+```csharp
+class Name
+{
+    public void Deconstruct(out string first, out string last) { first = First; last = Last; }
+    ...
+}
+// or
+static class Extensions
+{
+    public static void Deconstruct(this Name name, out string first, out string last) { first = name.First; last = name.Last; }
+}
+```
+
+Overload resolution for `Deconstruct` methods considers only the arity of the `Deconstruct` method. If multiple `Deconstruct` methods of the same arity are accessible, the expression is ambiguous and a binding time error occurs.
+
+If necessary to satisfy implicit conversions of the tuple member types, the compiler passes temporary variables to the `Deconstruct` method, instead of the ones declared in the deconstruction. For example, if `p` has
+
+```csharp
+void Deconstruct(out byte x, out byte y) ...;
+```
+
+The compiler translates
+
+```csharp
+(int x, int y) = p;
+```
+
+equivalently to:
+
+```csharp
+p.Deconstruct(out byte __x, out byte __y);
+(int x, int y) = (__x, __y);
+```
+
+### Additions to [Classes](../../spec/classes.md)
+
+> The following note should be added to the end of the section on [extension methods](../../spec/classes.md#extension-methods):
+
+> note: 
+>Extension methods on a tuple type apply to tuples with different element names:
+
```

Comment:
This is not normative.  It is a consequence of the identity conversions.

Reply:
Addressed. This note is not marked as informative.